### PR TITLE
Qemu driver

### PR DESCRIPTION
**Pull request by antoineduban**
The iaas part of nanocloud  is now built to accept different driver for different iaas, Making it easier to adapt to other iaas int the future. The first driver for a local qemu instance of windows is also in this pull request

### CODE REVIEW

File: nanocloud/routes/iaas/iaas.go [hunk 1]
```diff
@@ -1,60 +1,133 @@
...
+		if err != nil {
+			log.Error(err)
+		}
+	}
+
+	return c.JSON(http.StatusOK, hash{"data": res})
+}
+
+func StopVM(c *echo.Context) error {
+	machine, err := vms.Machine(c.Param("id"))
```
Comment (willrnch) [inline nanocloud/routes/iaas/iaas.go:114]: error check is missing here

File: nanocloud/routes/iaas/iaas.go [hunk 2]
```diff
@@ -1,60 +1,133 @@
...
-)
-
-func proxy(c *echo.Context) error {
-	r := c.Request()
-	path := r.URL.Path
+type hash map[string]interface{}
 
-	var resp *http.Response
-	var err error
+type JsonMachine struct {
```
Comment (willrnch) [inline nanocloud/routes/iaas/iaas.go]: This type doesn't need to be public

File: nanocloud/routes/iaas/iaas.go [hunk 3]
```diff
@@ -1,60 +1,133 @@
...
+		return c.JSON(
+			http.StatusInternalServerError, hash{
+				"errors": [1]hash{
+					hash{
+						"detail": err.Error(),
+					},
+				},
+			})
 	}
+	type Attr struct {
```
Comment (willrnch) [inline nanocloud/routes/iaas/iaas.go]: this type doesn't need to be public

File: nanocloud/routes/iaas/iaas.go [hunk 4]
```diff
@@ -1,60 +1,133 @@
...
+		Name   string `json:"name"`
+		Ip     string `json:"ip"`
+		Status string `json:"status"`
+		Id     string `json:"id"`
+	}
+	type VM struct {
```
Comment (willrnch) [inline nanocloud/routes/iaas/iaas.go]: this type doesn't need to be public

File: nanocloud/routes/iaas/iaas.go [hunk 5]
```diff
@@ -1,60 +1,133 @@
...
+	Status string
+	Ip     string
+}
 
-	if r.Method == "GET" {
-		resp, err = http.Get(iaasAPIurl + path)
-	} else {
-		resp, err = http.Post(iaasAPIurl+path, "", nil)
-	}
+func MachinetoStruct(rawmachine vm.Machine) JsonMachine {
```
Comment (willrnch) [inline nanocloud/routes/iaas/iaas.go]: This fun doesn't need to be public.

File: nanocloud/routes/iaas/iaas.go [hunk 6]
```diff
@@ -1,60 +1,133 @@
...
 	}
+	return c.JSON(
+		http.StatusOK, hash{
+			"vm": MachinetoStruct(machine),
+		})
+}
 
-	contentType := ""
+func StartVM(c *echo.Context) error {
+	machine, err := vms.Machine(c.Param("id"))
```
Comment (willrnch) [inline nanocloud/routes/iaas/iaas.go:130]: error check is missing

File: nanocloud/routes/iaas/iaas.go [hunk 7]
```diff
@@ -1,60 +1,133 @@
...
+		return c.JSON(
+			http.StatusInternalServerError, hash{
+				"errors": [1]hash{
+					hash{
+						"detail": err.Error(),
+					},
+				},
+			})
+	}
+	return c.JSON(
```
Comment (willrnch) [inline nanocloud/routes/iaas/iaas.go:147]: do not return 200 if `err != nil`

File: nanocloud/vms/drivers/qemu/machine.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+	if err != nil {
+		log.Error(err)
+		return vms.StatusUnknown, err
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.Error(err)
+		return vms.StatusUnknown, err
+	}
+	type Status struct {
```
Comment (Costent) [inline nanocloud/vms/drivers/qemu/machine.go]: useless

File: nanocloud/vms/drivers/qemu/machine.go [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+	resp, err := http.Get("http://" + string(ip) + ":8080/api/iaas")
+	if err != nil {
+		log.Error(err)
+		return vms.StatusUnknown, err
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.Error(err)
+		return vms.StatusUnknown, err
+	}
```
Comment (willrnch) [inline nanocloud/vms/drivers/qemu/machine.go:61]:
``` go
    var state struct {
        Data []struct {
            Id         string `json:"id"`
            Type       string `json:"type"`
            Attributes string
        } `json:"data"`
    }
```

File: nanocloud/vms/drivers/qemu/machine.go [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+		case "available":
+			return vms.StatusDown, nil
+		}
+		return vms.StatusUnknown, nil
+	}
+	return vms.StatusUnknown, nil
+}
+
+func (m *machine) IP() (net.IP, error) {
+	iaas := os.Getenv("WIN_SERVER")
```
Comment (willrnch) [inline nanocloud/vms/drivers/qemu/machine.go]: probably `EXECUTION_SERVER`

Comment (willrnch) [inline nanocloud/vms/drivers/qemu/machine.go]: Get this information from `open` options arguments.

File: nanocloud/vms/vms.go [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+	if drivers == nil {
+		drivers = make(map[string]Driver, 0)
+	}
+	drivers[name] = driver
+}
+
+func Open(driverName string, options map[string]string) (*VM, error) {
+	driver := drivers[driverName]
+	if driver != nil {
+		lala, _ := driver.Open(options)
```
Comment (willrnch) [inline nanocloud/vms/vms.go]: lala

File: nanocloud/vms/drivers/qemu/type.go [hunk 12]
```diff
@@ -0,0 +1,29 @@
+package qemu
```
Comment (willrnch) [inline nanocloud/vms/drivers/qemu/type.go:23]: missing licensing header

File: nanocloud/vms/drivers/qemu/vm.go [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+	resp, err := http.Post("http://"+string(ip)+":8080/api/iaas/"+m.Id()+"/download", "", nil)
+	if err != nil || resp.StatusCode != http.StatusOK {
+		log.Error(err)
+		return nil, err
+	}
+	return &m, nil
+}
+
+func (v *vm) Machines() ([]vms.Machine, error) {
+	iaas := os.Getenv("WIN_SERVER")
```
Comment (willrnch) [inline nanocloud/vms/drivers/qemu/vm.go]: Get this information from open options arguments.

**Pull request merged**